x/upgrade/keeper/grpc_query.go

```diff
@@ -46,3 +48,35 @@ func (k Keeper) UpgradedConsensusState(c context.Context, req *types.QueryUpgrad
 		UpgradedConsensusState: consState,
 	}, nil
 }
+
+// ModuleVersions implements the Query/QueryModuleVersions gRPC method
+func (k Keeper) ModuleVersions(c context.Context, req *types.QueryModuleVersionsRequest) (*types.QueryModuleVersionsResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	// get version map from x/upgrade store
+	vm := k.GetModuleVersionMap(ctx)
+
+	// make response slice
+	res := make([]*types.ModuleVersion, 0)
+
+	// check if a specific module was requested
+	if len(req.ModuleName) > 0 {
+		// check if the requested module exists
+		if version, found := vm[req.ModuleName]; found {
+			// add the requested module
+			res = append(res, &types.ModuleVersion{Name: req.ModuleName, Version: version})
+		} else { // module was requested, but not found
+			return &types.QueryModuleVersionsResponse{}, errors.Wrapf(errors.ErrNotFound, "x/upgrade: QueryModuleVersions")
+		}
+	} else {
+		// if no module requested, add entire vm to slice
+		for m, v := range vm {
+			res = append(res, &types.ModuleVersion{Name: m, Version: v})
+		}
+	}
+
+	res = xp.Sort(res)
```

Comment:
I'm wondering if sort is really needed. The KVStoreIterator is deterministic. We could just have another private keeper function `getModuleVersions() []*types.ModuleVersion` that loops through the KVStoreIterator, and this grpc_query handler would just return that function.

I think keys are iterated in natural order, so we might even get alphabetical ordering for free.

Reply:
(added methods) [https://github.com/cosmos/cosmos-sdk/pull/9073/files#diff-33ff09992f1a3583e2aa8eb932c62c833d52e446488070ed958eb672eed8da3eR115-R149].

Yeah, that seemed to work. Also I had to make it an exported method `GetModuleVersions()` so I could use it in my tests. I also added a `getModuleVersion` as a helper method to get a version given a module name. [usage](https://github.com/cosmos/cosmos-sdk/pull/9073/files#diff-34674ec4d53c9bb830f28c5924f807fdb12e3755819de575bb2f66b03dcf33e8R57)